File: FMELibrary/Nation.cs
namespace FMELibrary
{
    /// <summary>
    /// Represents a nation with all its properties and attributes.
    /// </summary>
    public class Nation
    {
        /// <summary>
        /// Gets or sets the unique identifier for the nation.
        /// </summary>
        public int Uid { get; set; }

        /// <summary>
        /// Gets or sets the nation identifier.
        /// </summary>
        public short Id { get; set; }

        /// <summary>
        /// Gets or sets the name of the nation.
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// Gets or sets the terminator byte after the name.
        /// </summary>
        public byte Terminator1 { get; set; }

        /// <summary>
        /// Gets or sets the nationality descriptor.
        /// </summary>
        public string Nationality { get; set; }

        /// <summary>
        /// Gets or sets the terminator byte after the nationality.
        /// </summary>
        public byte Terminator2 { get; set; }

        /// <summary>
        /// Gets or sets the code name of the nation.
        /// </summary>
        public string CodeName { get; set; }

        /// <summary>
        /// Gets or sets the continent identifier.
        /// </summary>
        public short ContinentId { get; set; }

        /// <summary>
        /// Gets or sets the capital city identifier.
        /// </summary>
        public short CityId { get; set; }

        /// <summary>
        /// Gets or sets the main stadium identifier.
        /// </summary>
        public short StadiumId { get; set; }

        /// <summary>
        /// Gets or sets an unknown integer value.
        /// </summary>
        public int Unknown1 { get; set; }

        /// <summary>
        /// Gets or sets an unknown short value.
        /// </summary>
        public short Unknown2 { get; set; }

        /// <summary>
        /// Gets or sets an unknown byte value.
        /// </summary>
        public byte Unknown3 { get; set; }

        /// <summary>
        /// Gets or sets the array of languages spoken in the nation (language ID and proficiency level).
        /// </summary>
        public (short, byte)[] Languages { get; set; }

        /// <summary>
        /// Gets or sets whether the nation is active in the game (1 = active, 0 = inactive).
        /// </summary>
        public byte IsActive { get; set; }

        /// <summary>
        /// Gets or sets the first national color (nullable, only if IsActive is 1).
        /// </summary>
        public short? Color1 { get; set; }

        /// <summary>
        /// Gets or sets an unknown integer value (nullable, only if IsActive is 1).
        /// </summary>
        public int? Unknown5 { get; set; }

        /// <summary>
        /// Gets or sets the second national color (nullable, only if IsActive is 1).
        /// </summary>
        public short? Color2 { get; set; }

        /// <summary>
        /// Gets or sets an unknown byte value (nullable, only if IsActive is 1).
        /// </summary>
        public byte? Unknown6 { get; set; }

        /// <summary>
        /// Gets or sets an unknown short value (nullable, only if IsActive is 1).
        /// </summary>
        public short? Unknown7 { get; set; }

        /// <summary>
        /// Gets or sets an unknown byte value (nullable, only if IsActive is 1).
        /// </summary>
        public byte? Unknown8 { get; set; }

        /// <summary>
        /// Gets or sets whether the nation has a FIFA/UEFA ranking (nullable, only if IsActive is 1).
        /// </summary>
        public byte? IsRanked { get; set; }

        /// <summary>
        /// Gets or sets the FIFA/UEFA ranking position (nullable, only if IsActive is 1).
        /// </summary>
        public short? Ranking { get; set; }

        /// <summary>
        /// Gets or sets the ranking points (nullable, only if IsActive is 1).
        /// </summary>
        public short? Points { get; set; }

        /// <summary>
        /// Gets or sets an unknown short value (nullable, only if IsActive is 1).
        /// </summary>
        public short? Unknown9 { get; set; }

        /// <summary>
        /// Gets or sets the first set of coefficient values (only if IsActive is 1).
        /// </summary>
        public float[] Coefficients1 { get; set; } = [];

        /// <summary>
        /// Gets or sets unknown data (11 bytes, only if IsActive is 1).
        /// </summary>
        public byte[] Unknown10 { get; set; } = [];

        /// <summary>
        /// Gets or sets whether the nation has a second coefficient set (1 = has coefficients, 0 = no coefficients).
        /// </summary>
        public byte HasCoefficient2 { get; set; }

        /// <summary>
        /// Gets or sets unknown data (16 bytes, only if HasCoefficient2 is 1).
        /// </summary>
        public byte[] Unknown11 { get; set; } = [];

        /// <summary>
        /// Gets or sets an unknown byte value (nullable, only if HasCoefficient2 is 1).
        /// </summary>
        public byte? Unknown12 { get; set; }

        /// <summary>
        /// Gets or sets an unknown short value (nullable, only if HasCoefficient2 is 1).
        /// </summary>
        public short? Unknown13 { get; set; }

        /// <summary>
        /// Gets or sets the second set of coefficient values (only if HasCoefficient2 is 1).
        /// </summary>
        public float[] Coefficients2 { get; set; } = [];

        /// <summary>
        /// Gets or sets unknown data (11 bytes, only if HasCoefficient2 is 1).
        /// </summary>
        public byte[] Unknown14 { get; set; } = [];

        /// <summary>
        /// Initializes a new instance of the <see cref="Nation"/> class by reading from a binary reader.
        /// </summary>
        /// <param name="reader">The binary reader containing the nation data.</param>
        public Nation(BinaryReader reader)
        {
            Uid = reader.ReadInt32();
            Id = reader.ReadInt16();

            Name = reader.ReadStringEx();
            Terminator1 = reader.ReadByte();
            Nationality = reader.ReadStringEx();
            Terminator2 = reader.ReadByte();
            CodeName = reader.ReadStringEx();

            ContinentId = reader.ReadInt16();
            CityId = reader.ReadInt16();
            StadiumId = reader.ReadInt16();

            Unknown1 = reader.ReadInt32();
            Unknown2 = reader.ReadInt16();
            Unknown3 = reader.ReadByte();

            Languages = new (short, byte)[reader.ReadByte()];
            for (int i = 0; i < Languages.Length; i++)
                Languages[i] = (reader.ReadInt16(), reader.ReadByte());

            IsActive = reader.ReadByte();
            if (IsActive == 1)
            {
                Color1 = reader.ReadInt16();
                Unknown5 = reader.ReadInt32();
                Color2 = reader.ReadInt16();

                Unknown6 = reader.ReadByte();
                Unknown7 = reader.ReadInt16();
                Unknown8 = reader.ReadByte();

                IsRanked = reader.ReadByte();
                Ranking = reader.ReadInt16();
                Points = reader.ReadInt16();

                Unknown9 = reader.ReadInt16();

                Coefficients1 = new float[reader.ReadByte()];
                for (int i = 0; i < Coefficients1.Length; i++)
                    Coefficients1[i] = reader.ReadSingle();

                Unknown10 = reader.ReadBytes(11);
            }

            HasCoefficient2 = reader.ReadByte();
            if (HasCoefficient2 == 1)
            {
                Unknown11 = reader.ReadBytes(16);
                Unknown12 = reader.ReadByte();
                Unknown13 = reader.ReadInt16();

                Coefficients2 = new float[reader.ReadByte()];
                for (int i = 0; i < Coefficients2.Length; i++)
                    Coefficients2[i] = reader.ReadSingle();

                Unknown14 = reader.ReadBytes(11);
            }
        }

        /// <summary>
        /// Converts the nation data to a byte array.
        /// </summary>
        /// <returns>A byte array representing the serialized nation data.</returns>
        public byte[] ToBytes()
        {
            using var stream = new MemoryStream();
            using var writer = new BinaryWriter(stream);
            Write(writer);
            return stream.ToArray();
        }

        /// <summary>
        /// Writes the nation data to the specified binary writer.
        /// </summary>
        /// <param name="writer">The binary writer to write the nation data to.</param>
        public void Write(BinaryWriter writer)
        {
            writer.WriteEx(Uid);
            writer.WriteEx(Id);

            writer.WriteEx(Name);
            writer.WriteEx(Terminator1);
            writer.WriteEx(Nationality);
            writer.WriteEx(Terminator2);
            writer.WriteEx(CodeName);

            writer.WriteEx(ContinentId);
            writer.WriteEx(CityId);
            writer.WriteEx(StadiumId);

            writer.WriteEx(Unknown1);
            writer.WriteEx(Unknown2);
            writer.WriteEx(Unknown3);

            writer.WriteEx((byte)Languages.Length);
            for (int i = 0; i < Languages.Length; i++)
            {
                writer.WriteEx(Languages[i].Item1);
                writer.WriteEx(Languages[i].Item2);
            }

            writer.WriteEx(IsActive);

            if (IsActive == 1)
            {
                writer.WriteEx(Color1);
                writer.WriteEx(Unknown5);
                writer.WriteEx(Color2);

                writer.WriteEx(Unknown6);
                writer.WriteEx(Unknown7);
                writer.WriteEx(Unknown8);

                writer.WriteEx(IsRanked);
                writer.WriteEx(Ranking);
                writer.WriteEx(Points);

                writer.WriteEx(Unknown9);
                writer.WriteEx((byte)Coefficients1.Length);
                for (int i = 0; i < Coefficients1.Length; i++)
                {
                    writer.WriteEx(Coefficients1[i]);
                }

                writer.WriteEx(Unknown10);
            }

            writer.WriteEx(HasCoefficient2);
            if (HasCoefficient2 == 1)
            {
                writer.WriteEx(Unknown11);
                writer.WriteEx(Unknown12);
                writer.WriteEx(Unknown13);

                writer.WriteEx((byte)Coefficients2.Length);
                for (int i = 0; i < Coefficients2.Length; i++)
                {
                    writer.WriteEx(Coefficients2[i]);
                }

                writer.WriteEx(Unknown14);
            }

            //writer.Write((byte)ExtraNames.Length);
            //for (int i = 0; i < ExtraNames.Length; i++)
            //{
            //    writer.WriteEx(ExtraNames[i].Item1);
            //    writer.Write(ExtraNames[i].Item2);
            //    writer.Write(ExtraNames[i].Item3);
            //}
        }
    }
}

File: FMELibrary.js/Nation.ts
import { BinaryParser } from './BinaryParser';

/**
 * Represents a nation with all its properties and attributes.
 */
export class Nation {
  /** Gets or sets the unique identifier for the nation. */
  uid: number = 0;

  /** Gets or sets the nation identifier. */
  id: number = 0;

  /** Gets or sets the name of the nation. */
  name: string = '';

  /** Gets or sets the terminator byte after the name. */
  terminator1: number = 0;

  /** Gets or sets the nationality descriptor. */
  nationality: string = '';

  /** Gets or sets the terminator byte after the nationality. */
  terminator2: number = 0;

  /** Gets or sets the code name of the nation. */
  codeName: string = '';

  /** Gets or sets the continent identifier. */
  continentId: number = 0;

  /** Gets or sets the capital city identifier. */
  cityId: number = 0;

  /** Gets or sets the main stadium identifier. */
  stadiumId: number = 0;

  /** Gets or sets an unknown integer value. */
  unknown1: number = 0;

  /** Gets or sets an unknown short value. */
  unknown2: number = 0;

  /** Gets or sets an unknown byte value. */
  unknown3: number = 0;

  /** Gets or sets the array of languages spoken in the nation (language ID and proficiency level). */
  languages: { languageId: number; proficiency: number }[] = [];

  /** Gets or sets whether the nation is active in the game (1 = active, 0 = inactive). */
  isActive: number = 0;

  /** Gets or sets the first national color (nullable, only if isActive is 1). */
  color1?: number;

  /** Gets or sets an unknown integer value (nullable, only if isActive is 1). */
  unknown5?: number;

  /** Gets or sets the second national color (nullable, only if isActive is 1). */
  color2?: number;

  /** Gets or sets an unknown byte value (nullable, only if isActive is 1). */
  unknown6?: number;

  /** Gets or sets an unknown short value (nullable, only if isActive is 1). */
  unknown7?: number;

  /** Gets or sets an unknown byte value (nullable, only if isActive is 1). */
  unknown8?: number;

  /** Gets or sets whether the nation has a FIFA/UEFA ranking (nullable, only if isActive is 1). */
  isRanked?: number;

  /** Gets or sets the FIFA/UEFA ranking position (nullable, only if isActive is 1). */
  ranking?: number;

  /** Gets or sets the ranking points (nullable, only if isActive is 1). */
  points?: number;

  /** Gets or sets an unknown short value (nullable, only if isActive is 1). */
  unknown9?: number;

  /** Gets or sets the first set of coefficient values (only if isActive is 1). */
  coefficients1: number[] = [];

  /** Gets or sets unknown data (11 bytes, only if isActive is 1). */
  unknown10: Buffer = Buffer.alloc(0);

  /** Gets or sets whether the nation has a second coefficient set (1 = has coefficients, 0 = no coefficients). */
  hasCoefficient2: number = 0;

  /** Gets or sets unknown data (16 bytes, only if hasCoefficient2 is 1). */
  unknown11: Buffer = Buffer.alloc(0);

  /** Gets or sets an unknown byte value (nullable, only if hasCoefficient2 is 1). */
  unknown12?: number;

  /** Gets or sets an unknown short value (nullable, only if hasCoefficient2 is 1). */
  unknown13?: number;

  /** Gets or sets the second set of coefficient values (only if hasCoefficient2 is 1). */
  coefficients2: number[] = [];

  /** Gets or sets unknown data (11 bytes, only if hasCoefficient2 is 1). */
  unknown14: Buffer = Buffer.alloc(0);

  /**
   * Initializes a new instance of the Nation class by reading from a buffer.
   * @param buffer - The buffer containing the nation data
   * @param offset - The current offset in the buffer
   * @returns An object containing the Nation instance and the new offset
   */
  static fromBuffer(buffer: Buffer, offset: number): { nation: Nation; offset: number } {
    const nation = new Nation();

    nation.uid = buffer.readInt32LE(offset);
    offset += 4;

    nation.id = buffer.readInt16LE(offset);
    offset += 2;

    const nameResult = BinaryParser.readStringEx(buffer, offset);
    nation.name = nameResult.value;
    offset = nameResult.offset;

    nation.terminator1 = buffer.readUInt8(offset);
    offset += 1;

    const nationalityResult = BinaryParser.readStringEx(buffer, offset);
    nation.nationality = nationalityResult.value;
    offset = nationalityResult.offset;

    nation.terminator2 = buffer.readUInt8(offset);
    offset += 1;

    const codeNameResult = BinaryParser.readStringEx(buffer, offset);
    nation.codeName = codeNameResult.value;
    offset = codeNameResult.offset;

    nation.continentId = buffer.readInt16LE(offset);
    offset += 2;

    nation.cityId = buffer.readInt16LE(offset);
    offset += 2;

    nation.stadiumId = buffer.readInt16LE(offset);
    offset += 2;

    nation.unknown1 = buffer.readInt32LE(offset);
    offset += 4;

    nation.unknown2 = buffer.readInt16LE(offset);
    offset += 2;

    nation.unknown3 = buffer.readUInt8(offset);
    offset += 1;

    const languagesCount = buffer.readUInt8(offset);
    offset += 1;

    nation.languages = [];
    for (let i = 0; i < languagesCount; i++) {
      const languageId = buffer.readInt16LE(offset);
      offset += 2;
      const proficiency = buffer.readUInt8(offset);
      offset += 1;
      nation.languages.push({ languageId, proficiency });
    }

    nation.isActive = buffer.readUInt8(offset);
    offset += 1;

    if (nation.isActive === 1) {
      nation.color1 = buffer.readInt16LE(offset);
      offset += 2;

      nation.unknown5 = buffer.readInt32LE(offset);
      offset += 4;

      nation.color2 = buffer.readInt16LE(offset);
      offset += 2;

      nation.unknown6 = buffer.readUInt8(offset);
      offset += 1;

      nation.unknown7 = buffer.readInt16LE(offset);
      offset += 2;

      nation.unknown8 = buffer.readUInt8(offset);
      offset += 1;

      nation.isRanked = buffer.readUInt8(offset);
      offset += 1;

      nation.ranking = buffer.readInt16LE(offset);
      offset += 2;

      nation.points = buffer.readInt16LE(offset);
      offset += 2;

      nation.unknown9 = buffer.readInt16LE(offset);
      offset += 2;

      const coefficients1Count = buffer.readUInt8(offset);
      offset += 1;

      nation.coefficients1 = [];
      for (let i = 0; i < coefficients1Count; i++) {
        nation.coefficients1.push(buffer.readFloatLE(offset));
        offset += 4;
      }

      nation.unknown10 = buffer.slice(offset, offset + 11);
      offset += 11;
    }

    nation.hasCoefficient2 = buffer.readUInt8(offset);
    offset += 1;

    if (nation.hasCoefficient2 === 1) {
      nation.unknown11 = buffer.slice(offset, offset + 16);
      offset += 16;

      nation.unknown12 = buffer.readUInt8(offset);
      offset += 1;

      nation.unknown13 = buffer.readInt16LE(offset);
      offset += 2;

      const coefficients2Count = buffer.readUInt8(offset);
      offset += 1;

      nation.coefficients2 = [];
      for (let i = 0; i < coefficients2Count; i++) {
        nation.coefficients2.push(buffer.readFloatLE(offset));
        offset += 4;
      }

      nation.unknown14 = buffer.slice(offset, offset + 11);
      offset += 11;
    }

    return { nation, offset };
  }

  /**
   * Converts the nation data to a buffer.
   * @returns A buffer representing the serialized nation data
   */
  toBuffer(): Buffer {
    let size = 4 + 2; // uid + id
    size += 4 + Buffer.byteLength(this.name, 'utf-8'); // name
    size += 1; // terminator1
    size += 4 + Buffer.byteLength(this.nationality, 'utf-8'); // nationality
    size += 1; // terminator2
    size += 4 + Buffer.byteLength(this.codeName, 'utf-8'); // codeName
    size += 2 + 2 + 2; // continentId, cityId, stadiumId
    size += 4 + 2 + 1; // unknown1, unknown2, unknown3
    size += 1 + (this.languages.length * 3); // languages
    size += 1; // isActive

    if (this.isActive === 1) {
      size += 2 + 4 + 2 + 1 + 2 + 1 + 1 + 2 + 2 + 2; // colors and unknowns
      size += 1 + (this.coefficients1.length * 4); // coefficients1
      size += 11; // unknown10
    }

    size += 1; // hasCoefficient2

    if (this.hasCoefficient2 === 1) {
      size += 16 + 1 + 2; // unknown11, unknown12, unknown13
      size += 1 + (this.coefficients2.length * 4); // coefficients2
      size += 11; // unknown14
    }

    const buffer = Buffer.alloc(size);
    let offset = 0;

    buffer.writeInt32LE(this.uid, offset);
    offset += 4;

    buffer.writeInt16LE(this.id, offset);
    offset += 2;

    offset = BinaryParser.writeEx(this.name, buffer, offset);
    offset = BinaryParser.writeByte(this.terminator1, buffer, offset);
    offset = BinaryParser.writeEx(this.nationality, buffer, offset);
    offset = BinaryParser.writeByte(this.terminator2, buffer, offset);
    offset = BinaryParser.writeEx(this.codeName, buffer, offset);

    buffer.writeInt16LE(this.continentId, offset);
    offset += 2;

    buffer.writeInt16LE(this.cityId, offset);
    offset += 2;

    buffer.writeInt16LE(this.stadiumId, offset);
    offset += 2;

    buffer.writeInt32LE(this.unknown1, offset);
    offset += 4;

    buffer.writeInt16LE(this.unknown2, offset);
    offset += 2;

    buffer.writeUInt8(this.unknown3, offset);
    offset += 1;

    buffer.writeUInt8(this.languages.length, offset);
    offset += 1;

    for (const lang of this.languages) {
      buffer.writeInt16LE(lang.languageId, offset);
      offset += 2;
      buffer.writeUInt8(lang.proficiency, offset);
      offset += 1;
    }

    buffer.writeUInt8(this.isActive, offset);
    offset += 1;

    if (this.isActive === 1) {
      buffer.writeInt16LE(this.color1!, offset);
      offset += 2;

      buffer.writeInt32LE(this.unknown5!, offset);
      offset += 4;

      buffer.writeInt16LE(this.color2!, offset);
      offset += 2;

      buffer.writeUInt8(this.unknown6!, offset);
      offset += 1;

      buffer.writeInt16LE(this.unknown7!, offset);
      offset += 2;

      buffer.writeUInt8(this.unknown8!, offset);
      offset += 1;

      buffer.writeUInt8(this.isRanked!, offset);
      offset += 1;

      buffer.writeInt16LE(this.ranking!, offset);
      offset += 2;

      buffer.writeInt16LE(this.points!, offset);
      offset += 2;

      buffer.writeInt16LE(this.unknown9!, offset);
      offset += 2;

      buffer.writeUInt8(this.coefficients1.length, offset);
      offset += 1;

      for (const coef of this.coefficients1) {
        buffer.writeFloatLE(coef, offset);
        offset += 4;
      }

      this.unknown10.copy(buffer, offset);
      offset += 11;
    }

    buffer.writeUInt8(this.hasCoefficient2, offset);
    offset += 1;

    if (this.hasCoefficient2 === 1) {
      this.unknown11.copy(buffer, offset);
      offset += 16;

      buffer.writeUInt8(this.unknown12!, offset);
      offset += 1;

      buffer.writeInt16LE(this.unknown13!, offset);
      offset += 2;

      buffer.writeUInt8(this.coefficients2.length, offset);
      offset += 1;

      for (const coef of this.coefficients2) {
        buffer.writeFloatLE(coef, offset);
        offset += 4;
      }

      this.unknown14.copy(buffer, offset);
      offset += 11;
    }

    return buffer.slice(0, offset);
  }
}
